src/algorithms/synthesis/dd_synthesis.cpp

```diff
@@ -433,21 +428,83 @@ namespace syrec {
         return unifyPath(src, current, p1SigVec, p2SigVec, changePaths, dd);
     }
 
-    // This function returns the operations required to synthesize the DD.
-    auto DDSynthesizer::synthesize(dd::mEdge src, std::unique_ptr<dd::Package<>>& dd) -> qc::QuantumComputation& {
-        qc.clear();
-        runtime  = 0.;
-        numGates = 0U;
+    // Refer to the decoder algorithm of https://www.cda.cit.tum.de/files/eda/2018_aspdac_coding_techniques_in_synthesis.pdf.
+    auto DDSynthesizer::decoder(TruthTable::CubeMap const& codewords, std::unique_ptr<dd::Package<>>& dd) -> void {
+        const auto codeLength = codewords.begin()->second.size();
+
+        // decode the r most significant bits of the original output pattern.
+        if (r != 0U) {
+            for (const auto& [pattern, code]: codewords) {
+                TruthTable::Cube targetCube(pattern.begin(), pattern.begin() + static_cast<int>(r));
+
+                dd::Controls ctrl;
+                for (auto i = 0U; i < codeLength; i++) {
+                    if (code[i].has_value()) {
+                        const auto ctrlType = *code[i] ? dd::Control::Type::pos : dd::Control::Type::neg;
+                        ctrl.emplace(dd::Control{static_cast<dd::Qubit>((codeLength - 1U) - i), ctrlType});
```

Comment:
The `dd::Control` part here is superfluous

Reply:
May be a stupid question. I need to construct the `dd::Control` object in order to emplace in ctrl right?

Reply:
DO you mean I need to do this `ctrl.emplace(static_cast<dd::Qubit>((codeLength - 1U) - i), ctrlType)`?

Reply:
Interesting. And if you do `.emplace({...})`?

Reply:
same error

Reply:
Nevermind then.